const fs = require("fs");
const { promisify } = require("util");
const unlinkAsync = promisify(fs.unlink);
const multerUtils = require("./middleware/uploadProfile");
const multerUtil = require("./middleware/uploadRecipeImages")
const multer = require("multer");

const uploadSingle = (req, res, next) => {
  const uploadSingle = multerUtils.single("image", 1);

  uploadSingle(req, res, (err) => {
    try {
      if (err instanceof multer.MulterError) {
        // A Multer error occurred when uploading.
        res.status(400).send(err?.message ?? "Something went wrong!");
        return;
      } else if (err) {
        // An unknown error occurred when uploading.
        res.status(400).send(err ?? "Something went wrong!");
        return;
      }

      next();
    } catch (error) {
      console.log("error",error)
      res.status(500).send(error?.message ?? "Upload Failed");
    }
  });
};

const uploadMultiple = (req, res, next) => {
  const uploadSingle = multerUtils.array("images", 5);

  uploadSingle(req, res, (err) => {
    try {
      if (err instanceof multer.MulterError) {
        if (err.code === "LIMIT_UNEXPECTED_FILE"){
          res.status(400).send("Error: File cannot be more than 5");
          return;
        }
        // A Multer error occurred when uploading.
        res.status(400).send(err?.message ?? "Something went wrong!");
        return;
      } else if (err) {
        // An unknown error occurred when uploading.
        res.status(400).send(err ?? "Something went wrong!");
        return;
      }

      next();
      // if (err) throw err;
    } catch (error) {
      res.status(500).send(error?.message ?? "Upload Failed");
    }
  });
};

const deleteFile = async (req, res) => {
  await unlinkAsync(`images/${req.body.name}`);
  res.send("DELETE COMPLETED!");
};

module.exports = { deleteFile, uploadMultiple, uploadSingle };